docs/proposals/configmap-rollout.md

```diff
@@ -0,0 +1,181 @@
+deployment "bar" updated
+```
+
+A new controller needs to watch for Deployments and ConfigMaps. Whenever it
+spots a Deployment with a `triggered-by` annotation and a mounted ConfigMap,
+it will create a copy of the ConfigMap and mount the copy in the Deployment.
+It should generally track Deployments with `triggered-by` annotations, 
+create new ConfigMap copies on updates of the original ConfigMap, and mount
+them back in the Deployment. Only Deployments should work for this controller
+for now but in the future the controller can be extended to facilitate config 
+management for DaemonSets, PetSets, and any other resource that may need it.
+
+Once the copied ConfigMap is mounted in the Deployment, the deployment
+controller needs to create the new ReplicaSet and set the owner reference
+for the copy to point to the ReplicaSet. In that way we ensure that copies
+of the original ConfigMap will be pruned by the garbage collector once their
+owner ReplicaSets are deleted. Certain copies of a ConfigMap will end up 
+being shared by more than one ReplicaSets.
+
+Additionally, the deployment controller should not create a new ReplicaSet
+for a Deployment "bar" that holds a `triggered-by` annotation to a ConfigMap
+"foo". Instead, it should wait until the new controller updates the Deployment
+template with the first copy of "foo". This ensures that the original ConfigMap
+copy will not land in a ReplicaSet, hence will not be subject to garbage
+collection.
+
+
+### Alternative solutions
+
+* Embed ConfigMap as a template in the Deployment.
+
+While it will work smoothly on ConfigMap changes for Deployments, this
+solution will need to be spread across the API for all resources that face
+the same problem. Sooner or later, higher-level resources will get bloated
+trying to manage stuff like ConfigMaps, Secrets, et al.
+
+* Others?
+
+## Changes
+
+### API
+
+No API changes will be needed for Deployments or ConfigMaps.
+
+### Controller
+
+A new controller will be needed for watching Deployments and Config Maps. Changes will also be needed in the deployment controller. [See above](#proposed-solution).
+
+### CLI
+
+We need a way to specify relationship between a resource that is updated (ConfigMap in our case) and a resource that will be triggered on those updates (Deployment). A generic `set trigger` will be added in `kubectl`. The relationship will be established with an annotation on the triggered resource (Deployment).
+```sh
+$ kubectl set trigger deployment/bar configmap/foo
+deployment "bar" updated
+$ kubectl get deployment/bar -o yaml | grep triggered-by
+deployment.kubernetes.io/triggered-by: configmap/foo
+```
+
+We also need a way to set volumes in pod templates without having to edit files by hand. A `kubectl set volume` command will be added to accommodate that need.
+```sh
+$ kubectl set volume deployment/bar -t configmap --configmapname foo
+deployment "bar" updated
+```
+
+## Issues
```

Comment:
> If you don't provide the facility in a usable way at the kube level then I'm forced to figure out how to manage it on top of kube.

That's the problem here. If we can't find a usable way of offering automatic deployments on configuration changes, we may need to decide if we really want to do it on the Kube level or let users build on top (that's why we also want feedback on tools like the `observe` command: https://github.com/kubernetes/kubernetes/issues/5164). 

Did you try the PoC I have in place for this proposal? I would like some feedback on it. There seem to be two different cases for configuration changes between deployments. 1) You turn on a flag which means *only* ConfigMap changes. My PoC covers that case. 2) You rollout a new version of your app and turn on/off flags appropriately ie. both CM and Deployment changes. Bundling the CM into the Deployment would offer an atomic way of doing 2 but I am not sure that moving app configuration as part of the Deployment will be usable long-term. Eventually, trying to edit a mass JSON/YAML file is not something we want users to be doing. Maybe small scoped changes either on the Deployment or in app configuration (ConfigMap, Secret) and the ability to run one-off deployments is the best solution for Kubernetes to offer at this point.